presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java

```diff
@@ -197,20 +197,37 @@ public TaskStatus process()
 
     private void invokeNoMoreSplitsIfNecessary()
     {
-        if (partitions.isEmpty() && fileIterators.isEmpty()) {
-            taskExecutionLock.writeLock().lock();
-            try {
-                // the write lock guarantees that no one is operating on the partitions, fileIterators, or hiveSplitSource, or half way through doing so.
-                if (partitions.isEmpty() && fileIterators.isEmpty()) {
-                    // It is legal to call `noMoreSplits` multiple times or after `stop` was called.
-                    // Nothing bad will happen if `noMoreSplits` implementation calls methods that will try to obtain a read lock because the lock is re-entrant.
-                    hiveSplitSource.noMoreSplits();
-                }
+        taskExecutionLock.readLock().lock();
+        try {
+            if (!partitions.isEmpty() || !fileIterators.isEmpty()) {
+                return;
             }
-            finally {
-                taskExecutionLock.writeLock().unlock();
+        }
+        catch (Exception e) {
```

Comment:
The problem is as follows: partitions is lazily evaluated, the call to partitions.isEmpty() can throw, for the particular case I'm debugging is an offline partition exception. If here is no catch, the split source is not marked as failed and the query will hang. 
FYI, partitions uses com.google.common.collect.MultitransformedIterator, so during hasNext it will try to do a transformation on the object.